paramiko/ecdsakey.py

```diff
@@ -57,41 +116,53 @@ def __init__(self, msg=None, data=None, filename=None, password=None,
             msg = Message(data)
         if vals is not None:
             self.signing_key, self.verifying_key = vals
+            c_class = self.signing_key.curve.__class__
+            self.ecdsa_curve = self._ECDSA_CURVES.get_by_curve_class(c_class)
         else:
             if msg is None:
                 raise SSHException('Key object may not be empty')
-            if msg.get_text() != 'ecdsa-sha2-nistp256':
+            self.ecdsa_curve = self._ECDSA_CURVES.get_by_key_format_identifier(
+                msg.get_text())
+            if self.ecdsa_curve is None:
                 raise SSHException('Invalid key')
             curvename = msg.get_text()
-            if curvename != 'nistp256':
+            if curvename != self.ecdsa_curve.nist_name:
                 raise SSHException("Can't handle curve of type %s" % curvename)
 
             pointinfo = msg.get_binary()
             if pointinfo[0:1] != four_byte:
                 raise SSHException('Point compression is being used: %s' %
                                    binascii.hexlify(pointinfo))
-            curve = ec.SECP256R1()
+            curve = self.ecdsa_curve.curve_class()
+            key_bytes = (curve.key_size + 7) // 8
             numbers = ec.EllipticCurvePublicNumbers(
-                x=inflate_long(pointinfo[1:1 + curve.key_size // 8], always_positive=True),
-                y=inflate_long(pointinfo[1 + curve.key_size // 8:], always_positive=True),
+                x=inflate_long(pointinfo[1:1 + key_bytes],
+                               always_positive=True),
+                y=inflate_long(pointinfo[1 + key_bytes:],
+                               always_positive=True),
                 curve=curve
             )
```

Comment:
If your'e willing to depend on a recent pyca/cryptography, this mess can be replace with `ec.EllipcticCurvePublicNumbers.from_encoded_point(curve, pointinfo)`. I can send a PR to do that separately if you'd like.